src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUnboundIdentifiersDiagnosticAnalyzer.cs

```diff
@@ -46,19 +47,24 @@ protected override bool ConstructorDoesNotExist(SyntaxNode node, SymbolInfo info
             }
 
             var count = constructors.Value
-            .WhereAsArray(constructor => constructor.Parameters.Length == args.Count)
             .WhereAsArray(constructor =>
             {
-                for (var i = 0; i < constructor.Parameters.Length; i++)
+                if (constructor.Parameters.Length == args.Count)
                 {
-                    var typeInfo = model.GetTypeInfo(args[i].Expression);
-                    if (!constructor.Parameters[i].Type.Equals(typeInfo.ConvertedType))
-                    {
-                        return false;
-                    }
+                    return true;
                 }
+                else
+                {
+                    var optionalCount = GetCountOfOptionalParameters(constructor);
+
+                    return optionalCount + args.Count == constructor.Parameters.Length;
+                }
+            })
+            .WhereAsArray(constructor =>
+            {
+                var argToParameterDictionary = ArgumentToParameterMapping(args, constructor.Parameters, model);
+                return GetCountOfOptionalParameters(constructor) + argToParameterDictionary.Count == constructor.Parameters.Length;
 
-                return true;
             }).Length;
```

Comment:
yo have two .WhereAsArray in a row.  can we just have one, that calls into helper methods that do whatever work you're tryign to do here?

Reply:
tbh, i have zero clue what this code is doing, or why.  why we do need to report our own diagnostics here instead of having the compiler do it?

Reply:
https://github.com/dotnet/roslyn/issues/1867
Apparently the compiler does not return semantic errors inside lambdas if there are syntax errors.

Reply:
Unless this has been fixed and the analyzer is no longer necessary?

Reply:
@akhera99 This is tracked by https://github.com/dotnet/roslyn/issues/32628.